Configure serialization of metadata per MessageVerifier object

etiennebarrie opened the pull request:
Follow up to #46848 cc @jonathanhefner 

### Motivation / Background

In our application, we use some message verifiers to sign messages sent to other Rails applications. The application runs on main but not all the other applications do.

We'd like to be able to configure the serialization per message verifier object, that way we can avoid the double serialization of message and metadata for some verifiers (those used by the application itself, or by other applications running main), but leave the serialization compatible for other verifiers.

### Detail

Instead of directly using the singleton `ActiveSupport::Messages::Metadata.use_message_serializer_for_metadata`, each Codec object has its own instance variable, the value of which is taken whenever the Codec object is instantiated. That could lead to issues if a Codec object is instantiated before the configuration is copied from the application config to the singleton, which happens in a `after_initialize` block, so after the initializers have run. I don't know how often that happens, should we handle that differently by using a sentinel value like `:default`, and checking the configuration at the last possible moment instead?

### Additional information

I didn't add a changelog entry because #46848 didn't and I'm not so sure how to word it.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why.
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature.

etiennebarrie committed:
Configure serialization of metadata per MessageVerifier object

jonathanhefner commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -128,9 +138,9 @@ class InvalidSignature < StandardError; end
 
     cattr_accessor :default_message_verifier_serializer, instance_accessor: false, default: :marshal
 
-    def initialize(secret, digest: nil, serializer: nil, url_safe: false)
+    def initialize(secret, digest: nil, **codec_options)
```
I know that we already mention the `:serializer` and `:url_safe` options in the docs, but they aren't very easy to scan, in my opinion.  If we remove `:serializer` and `:url_safe` from the method signature, then perhaps we should add an `=== Options` section to the documentation to increase their visibility.  (It could replace the `=== Generating URL-safe strings` and `=== Backwards compatibility` sections.)

If you decide to go that way, I think a dictionary-style list makes it easier to include code examples ([like so](https://github.com/rails/rails/blob/2675c906b1cc4cc877ddee4d62dc33f544adc748/actionpack/lib/action_controller/metal/rendering.rb#L98-L134)).

etiennebarrie commented:
Ok I brought back the keyword arguments, changed the documentation to use the dictionary-style list, changed the implementation to avoid breaking the encapsulation. I went with `force_legacy_metadata_serializer` instead of `force_default_metadata_serializer` because I felt it was unclear what default it was referring to.

jonathanhefner reviewed (commented):
Looks great!  Small typo, but I am pushing a commit to fix it.  Will merge on green.

jonathanhefner commented:
Thank you, @etiennebarrie! :envelope:
